debezium-server/debezium-server-rocketmq/src/main/java/io/debezium/server/rocketmq/RocketMqChangeConsumer.java

```diff
@@ -0,0 +1,191 @@
+    }
+
+    @PreDestroy
+    void close() {
+        // Closed rocketmq producer
+        LOGGER.info("Consumer destroy...");
+        if (mqProducer != null) {
+            mqProducer.shutdown();
+        }
+    }
+
+    @Override
+    public void handleBatch(List<ChangeEvent<Object, Object>> records, DebeziumEngine.RecordCommitter<ChangeEvent<Object, Object>> committer)
+            throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(records.size());
+        Deque<SubmittedRecord> submittedRecords = new LinkedList<>();
+        for (ChangeEvent<Object, Object> record : records) {
+            try {
+                final String topicName = streamNameMapper.map(record.destination());
+                SubmittedRecord submittedRecord = new SubmittedRecord(record);
+                String key = getString(record.key());
+                mqProducer.send(new Message(topicName, null, key, getBytes(record.value())), new SelectMessageQueueByHash(), key, new SendCallback() {
+                    @Override
+                    public void onSuccess(SendResult sendResult) {
+                        LOGGER.debug("Sent message with offset: {}", sendResult.getQueueOffset());
+                        submittedRecord.ack(latch);
+                    }
+
+                    @Override
+                    public void onException(Throwable throwable) {
+                        LOGGER.error("Failed to send record to {}:", record.destination(), throwable);
+                        throw new DebeziumException(throwable);
+                    }
+                });
+                submittedRecords.add(submittedRecord);
+            } catch (Exception e) {
+                // Ensure that the data has been sent successfully and is marked as processed
+                internalMarkProcessed(submittedRecords, committer);
+                throw new DebeziumException(e);
+            }
+        }
+        latch.await();
+        // Mark processed
+        internalMarkProcessed(submittedRecords, committer);
+        committer.markBatchFinished();
+    }
+
+    public void internalMarkProcessed(Deque<SubmittedRecord> submittedRecords, DebeziumEngine.RecordCommitter<ChangeEvent<Object, Object>> committer)
+            throws InterruptedException {
+        long totalDataSubmitted = submittedRecords.size();
+        LOGGER.debug("{} records will be marked as processed", submittedRecords.size());
+        if (submittedRecords == null || submittedRecords.peek() == null) {
+            LOGGER.warn("No records to mark processed!");
+            return;
+        }
+        long markProcessedCount = 0;
+        while (submittedRecords.peek() != null && submittedRecords.peek().getAcked().get()) {
+            committer.markProcessed(submittedRecords.poll().getRecord());
+            ++markProcessedCount;
+        }
+
+        if (submittedRecords.peek() != null && !submittedRecords.peek().getAcked().get()) {
```

Comment:
If you decide to wait for all event, this IMHO cannot happen as message is either delivered (ack is true) or exception is throws before you get to this point (you will be still waiting in `latch.await();`). Or am I missing something?

Reply:
> If you decide to wait for all event, this IMHO cannot happen as message is either delivered (ack is true) or exception is throws before you get to this point (you will be still waiting in `latch.await();`). Or am I missing something?

I think it is difficult to control the timeout. So I think it's enough to throw an exception and let us observe it

Reply:
I'm really no sure we need it - your original version was just fine (besides wrong place of `committer.markProcessed()`), I was missing the point that async message has default timeout for sending, so that `latch.await()` won't block forever. So IMHO this would work well:

```
// messages have set default timeout set to send.msg.timeout, so this cannot block forever
latch.await()

for (ChangeEvent<Object, Object> record : records) {
    committer.markProcessed(record);
}
```

and there's no need to copy the records to `SubmittedRecord` and the code more complicated.

Reply:
> I'm really no sure we need it - your original version was just fine (besides wrong place of `committer.markProcessed()`), I was missing the point that async message has default timeout for sending, so that `latch.await()` won't block forever. So IMHO this would work well:
> 
> ```
> // messages have set default timeout set to send.msg.timeout, so this cannot block forever
> latch.await()
> 
> for (ChangeEvent<Object, Object> record : records) {
>     committer.markProcessed(record);
> }
> ```
> 
> and there's no need to copy the records to `SubmittedRecord` and the code more complicated.

If there are [1,2,3,4] 4 records, we have successfully sent 1,2,4, and 3 has failed to submit, do we just need to mark that it has been processed to 2? The new queue is to ensure that the data of 3 will not be missed in the next startup

Reply:
AFAICT `markProcessed()` just updates in-memory state, actual flush is done in `markBatchFinished()`. So if you fail to deliver message number 3, you will throw an exception before flushing the offset to the disk, in-memory state is lost and next time will start from 1 again (stored offset), so nothing is lost. You get few duplicated messages, but this is in line with Debezium at-least-one delivery promise. I.e. if you fail, it IMHO doesn't matter much if you called `markProcessed()` or not (but I'd be very carefull when it's called in success case - it should be called only when we are sure message was delivered).

Reply:
> AFAICT `markProcessed()` just updates in-memory state, actual flush is done in `markBatchFinished()`. So if you fail to deliver message number 3, you will throw an exception before flushing the offset to the disk, in-memory state is lost and next time will start from 1 again (stored offset), so nothing is lost. You get few duplicated messages, but this is in line with Debezium at-least-one delivery promise. I.e. if you fail, it IMHO doesn't matter much if you called `markProcessed()` or not (but I'd be very carefull when it's called in success case - it should be called only when we are sure message was delivered).

According to this logic, my original logic is correct and does not need to be modified; The current way is to narrow the scope of duplicate data

Reply:
> AFAICT `markProcessed()` just updates in-memory state, actual flush is done in `markBatchFinished()`. So if you fail to deliver message number 3, you will throw an exception before flushing the offset to the disk, in-memory state is lost and next time will start from 1 again (stored offset), so nothing is lost. You get few duplicated messages, but this is in line with Debezium at-least-one delivery promise. I.e. if you fail, it IMHO doesn't matter much if you called `markProcessed()` or not (but I'd be very carefull when it's called in success case - it should be called only when we are sure message was delivered).

markProcessed() is only updated in memory status, but it also needs to be updated in order, because the same partition will be overwritten in memory. Moving markProcessed() to onSuccess() will cause disorder, and the final offset may not be what we want

Reply:
I already mentioned, that moving `markProcessed()` to `onSuccess()` was wrong proposal and you are right that it will mess up the order. But I still believe that you should `markProcessed()` only when you are sure that the messages was delivered, i.e. after `latch.await()`